qualtran/bloqs/arithmetic/addition.py

```diff
@@ -490,23 +431,46 @@ def build_composite_bloq(
             ctrls = None
         k = bb.allocate(dtype=x.reg.dtype)
 
-        xor_k_bloq = XorK(x.reg.dtype, self.k, self.cvs)
-        if ctrls is not None:
-            ctrls, k = bb.add(xor_k_bloq, ctrls=ctrls, x=k)
+        # Get binary representation of k and split k into separate wires.
+        k_split = bb.split(k)
+        if self.signed:
+            binary_rep = QInt(self.bitsize).to_bits(self.k)
         else:
-            k = bb.add(xor_k_bloq, x=k)
+            binary_rep = QUInt(self.bitsize).to_bits(self.k)
+
+        # Apply XGates to qubits in k where the bitstring has value 1. Apply CNOTs when the gate is
+        # controlled.
+        for i in range(self.bitsize):
+            if binary_rep[i] == 1:
+                if len(self.cvs) > 0 and ctrls is not None:
+                    ctrls, k_split[i] = bb.add(
+                        MultiControlX(cvs=self.cvs), ctrls=ctrls, x=k_split[i]
+                    )
+                else:
+                    k_split[i] = bb.add(XGate(), q=k_split[i])
```

Comment:
Sorry, why are we doing this change here instead of using `XorK` ? I think this is added after my last review?

Reply:
nvm, I got confused due to "show changes since your last review"